File: 백준/Silver/11048. 이동하기/이동하기.py
import sys
n, m  = map(int, sys.stdin.readline().split())
candy = []
dp = [[0 for _ in range(m)] for _ in range(n)]
for _ in range(n):
    candy.append(list(map(int, sys.stdin.readline().split())))

def dynamic():
    for i in range(n):
        for j in range(m):
            if i == j == 0:
                continue
            if i > 0 and j > 0:
                dp[i][j] = candy[i][j] + max(dp[i-1][j], dp[i][j-1])
            elif i > 0:
                dp[i][j] = candy[i][j] + dp[i-1][j]
            elif j > 0:
                dp[i][j] = candy[i][j] + dp[i][j-1]
    return dp[n-1][m-1]

dp[0][0] = candy[0][0]
print(dynamic())

File: 백준/Silver/11048. 이동하기/이동하기.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st;

		st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());

		int[][] miro = new int[N][M];

		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());

			for (int j = 0; j < M; j++) {
				miro[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		dynamic(N, M, miro);
	}

	private static void dynamic(int N, int M, int[][] miro) {

		int[][] dp = new int[N][M];
		dp[0][0] = miro[0][0];

		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				if (i + 1 < N) { // 아래로 이동 가능
					dp[i+1][j] = Math.max(dp[i][j] + miro[i+1][j], dp[i+1][j]);
				}

				if (i + 1 < N && j + 1 < M) {// 대각선으로 이동 가능
					dp[i+1][j+1] = Math.max(dp[i][j] + miro[i+1][j+1], dp[i+1][j+1]);
				}

				if (j + 1 < M) {
					dp[i][j+1] = Math.max(dp[i][j] + miro[i][j+1], dp[i][j+1]);
				}
			}
		}

		System.out.println(dp[N-1][M-1]);
	}
}
